content/posts/api-deployment-strategies.md

```diff
@@ -0,0 +1,341 @@
+{
+  "uri": "/v1/hello",
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "v1:80": 1
+    }
+  }
+}'
+```
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/2 -X PUT -d '
+{
+  "uri": "/v2/hello",
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "v2:80": 1
+    }
+  }
+}'
+```
+
+This configuration will route requests to path `/v1/hello` to the `v1` service and the requests to the `v2/hello` path to the `v2` service.
+
+### In Query Parameters
+
+Here, APISIX is configured to route traffic based on the version specified in the query parameter `version`. If the parameter has a value `v1`, the request is routed to the `v1` service, and if it has the value `v2`, the request is routed to the `v2` service:
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
+{
+  "uri": "/hello",
+  "vars": [["arg_version", "==", "v1"]],
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "V1:80": 1
+    }
+  }
+}'
+```
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/2 -X PUT -d '
+{
+  "uri": "/hello",
+  "vars": [["arg_version", "==", "v2"]],
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "V2:80": 1
+    }
+  }
+}'
+```
+
+### In Request Headers
+
+Similar to routing based on query parameters, the below configuration routes requests based on headers. While `arg_<argument>` checks for `<argument>` in the query parameters, `http_<header>` checks for the `<header>` in the request headers:
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
+{
+  "uri": "/hello",
+  "vars": [["http_version", "==", "v1"]],
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "v1:80": 1
+    }
+  }
+}'
+```
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/2 -X PUT -d '
+{
+  "uri": "/hello",
+  "vars": [["http_version", "==", "v2"]],
+  "upstream": {
+    "type": "roundrobin",
+    "nodes": {
+      "v2:80": 1
+    }
+  }
+}'
+```
+
+## Backwards Compatible APIs
+
+Multiple API versions cannot always get you far and sometimes, change is inevitable. A strategy to make changes to your API without breaking your client applications is to ensure backwards compatibility. i.e., the interface between your clients and the API remains the same while your API changes.
+
+For example, if you change your API endpoint (from `/old/api` to `/new/api`), an API gateway can redirect client requests to the new API endpoint without changes to the client applications.
+
+APISIX implements this through the [redirect](https://apisix.apache.org/docs/apisix/plugins/redirect/) and the [proxy-rewrite](https://apisix.apache.org/docs/apisix/plugins/proxy-rewrite/) plugins. The example below shows how you can configure redirects:
+
+```shell
+curl http://127.0.0.1:9180/apisix/admin/routes/1 -X PUT -d '
+{
+  "uri": "/old/api/hello",
+  "plugins": {
+    "redirect": {
+      "uri": "/new/api/hello",
+      "ret_code": 301
+    }
+  }
+}'
+```
+
```

Comment:
Beware! You make it sound easier than it is.

It only works if the client is configured with automatic redirects, _e.g._, `curl -L`, and then you still will execute twice the number of requests.

It deserves at least a mention, if not a red blinking alert